src/libstd/thread/mod.rs

```diff
@@ -525,6 +526,35 @@ pub fn park_timeout(dur: Duration) {
 }
 
 ////////////////////////////////////////////////////////////////////////////////
+// ThreadId
+////////////////////////////////////////////////////////////////////////////////
+
+/// A unique identifier for a running thread.
+///
+/// A `ThreadId` is an opaque object that has a unique value for each thread
+/// that creates one. `ThreadId`s do not correspond to a thread's system-
+/// designated identifier.
+#[unstable(feature = "thread_id", issue = "21507")]
+#[derive(Eq, PartialEq, Copy, Clone)]
+pub struct ThreadId(usize);
+
+impl ThreadId {
+    /// Returns an identifier unique to the current calling thread.
+    #[unstable(feature = "thread_id", issue = "21507")]
+    pub fn current() -> ThreadId {
+        static THREAD_ID_COUNT: AtomicUsize = ATOMIC_USIZE_INIT;
+        #[thread_local] static mut THREAD_ID: ThreadId = ThreadId(0);
+
+        unsafe {
+            if THREAD_ID.0 == 0 {
+                THREAD_ID.0 = 1 + THREAD_ID_COUNT.fetch_add(1, Ordering::SeqCst);
+            }
+            THREAD_ID
```

Comment:
That's the only possibility for collision that I thought of. Would it be better to use a u64? If the number of threads you are creating is overflowing a 64 bits (highly unlikely, IMO), then there's not much we can really do, is there? What possible representation could we even use for thread IDs in that case? I can't imagine an OS being able to even handle up to u64-1 threads.